src/main/java/com/google/devtools/build/lib/server/GrpcServerImpl.java

```diff
@@ -451,6 +452,11 @@ public void serve() throws AbruptExitException {
     // and if that fails, try again with IPv4.
     InetSocketAddress address = new InetSocketAddress("[::1]", port);
     try {
+      // TODO(bazel-team): Remove the following check after upgrading netty to a version with a fix
+      //  for https://github.com/netty/netty/issues/10402
+      if (NetUtil.isIpV4StackPreferred()) {
+        throw new IOException("ipv4 is preferred on the system.");
```

Comment:
Throwing an exception is a little bit awkward.

Consider rearranging the code, something like:
```
InetSocketAddress address = null;

try {
  if (!ipv4preferred) {
    address = new InetSocketAddress("[::1]", port);
    server = ....
  }
} catch (IOException e) {
  address = null;
}

if (address == null) {
  // insert the code from the catch block here
}
```

Reply:
Hmm, that will cause the `ipv6Exception` to null when ipv4preferred is true, which is required for constructing the error message. 

Considering we probably could revert this when the bug is fixed in netty, can we keep the code simple like this?

Reply:
Fair point.

Reply:
Thanks!